internal/audit/kafka/kafka_test.go

```diff
@@ -0,0 +1,201 @@
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"strconv"
+	"testing"
+	"time"
+
+	"github.com/ory/dockertest/v3"
+	"github.com/ory/dockertest/v3/docker"
+	"github.com/stretchr/testify/require"
+	"github.com/twmb/franz-go/pkg/kadm"
+	"github.com/twmb/franz-go/pkg/kgo"
+
+	auditv1 "github.com/cerbos/cerbos/api/genpb/cerbos/audit/v1"
+	"github.com/cerbos/cerbos/internal/audit"
+	_ "github.com/cerbos/cerbos/internal/audit/kafka"
+	"github.com/cerbos/cerbos/internal/config"
+)
+
+const (
+	redpandaImage   = "redpandadata/redpanda"
+	redpandaVersion = "v23.1.5"
+
+	defaultIntegrationTopic = "cerbos"
+)
+
+func TestSyncProduce(t *testing.T) {
+	t.Parallel()
+
+	ctx := context.Background()
+
+	// setup kafka
+	uri := newKafkaBroker(t, defaultIntegrationTopic)
+	log, err := newLog(map[string]any{
+		"audit": map[string]any{
+			"enabled": true,
+			"backend": "kafka",
+			"kafka": map[string]any{
+				"brokers":     []string{uri},
+				"topic":       defaultIntegrationTopic,
+				"produceSync": true,
+			},
+		},
+	})
+	require.NoError(t, err)
+
+	// write audit log entries
+	err = log.WriteAccessLogEntry(ctx, func() (*auditv1.AccessLogEntry, error) {
+		return &auditv1.AccessLogEntry{
+			CallId: "01ARZ3NDEKTSV4RRFFQ69G5FA1",
+		}, nil
+	})
+	require.NoError(t, err)
+
+	err = log.WriteDecisionLogEntry(ctx, func() (*auditv1.DecisionLogEntry, error) {
+		return &auditv1.DecisionLogEntry{
+			CallId: "01ARZ3NDEKTSV4RRFFQ69G5FA2",
+		}, nil
+	})
+	require.NoError(t, err)
+
+	// validate we see this entries in kafka
+	records, err := fetchKafkaTopic(uri, defaultIntegrationTopic)
+	require.NoError(t, err)
+	require.Len(t, records, 2, "unexpected number of published audit log entries")
+}
+
+func TestAsyncProduce(t *testing.T) {
+	t.Parallel()
+
+	ctx := context.Background()
+
+	// setup kafka
+	uri := newKafkaBroker(t, defaultIntegrationTopic)
+	log, err := newLog(map[string]any{
+		"audit": map[string]any{
+			"enabled": true,
+			"backend": "kafka",
+			"kafka": map[string]any{
+				"brokers":     []string{uri},
+				"topic":       defaultIntegrationTopic,
+				"produceSync": false,
+			},
+		},
+	})
+	require.NoError(t, err)
+
+	// write audit log entries
+	err = log.WriteAccessLogEntry(ctx, func() (*auditv1.AccessLogEntry, error) {
+		return &auditv1.AccessLogEntry{
+			CallId: "01ARZ3NDEKTSV4RRFFQ69G5FA1",
+		}, nil
+	})
+	require.NoError(t, err)
+
+	err = log.WriteDecisionLogEntry(ctx, func() (*auditv1.DecisionLogEntry, error) {
+		return &auditv1.DecisionLogEntry{
+			CallId: "01ARZ3NDEKTSV4RRFFQ69G5FA2",
+		}, nil
+	})
+	require.NoError(t, err)
+
+	// validate we see this entries in kafka, eventually
+	require.Eventually(t, func() bool {
+		records, err := fetchKafkaTopic(uri, defaultIntegrationTopic)
+		require.NoError(t, err)
+		return len(records) == 2
+	}, 10*time.Second, 100*time.Millisecond, "expected to see audit log entries in kafka")
+}
+
+func newKafkaBroker(t *testing.T, topic string) string {
+	t.Helper()
+
+	hostPort, err := freePort()
```

Comment:
You can use [`util.GetFreePort()`](https://github.com/cerbos/cerbos/blob/583155638574716a139bb272950d96f9f0b26179/internal/util/network.go#L52)

However, in this case, I don't think you even need to figure the free port yourself. You can get the random port assigned to the container by using `resource.GetPort("9092/tcp")`. You don't have to do the host port mapping on the container using `PortBindings` either. :)

Reply:
I tried that, but I struggled to get it working with the advertised port & Kafka admin client

Reply:
Ah OK. I reckon it's the Kafka client trying to be clever and discover the brokers. That's fine then. Do you mind switching to `util.GetFreePort`?

Reply:
Yeah, that's what it looked to be. On connect, the metadata back from the broker was using the local address which was not the same outside the container. 

Will migrate to that util function 👍🏻